File: src/main/java/org/example/kalkulationsprogramm/domain/ZeitbuchungAudit.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Unveränderliches Audit-Protokoll für Zeitbuchungen.
 * Speichert Snapshots bei jeder Änderung für GoBD-Konformität.
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "zeitbuchung_audit", uniqueConstraints = @UniqueConstraint(name = "uk_zeitbuchung_audit_version", columnNames = {
                "zeitbuchung_id", "version" }))
public class ZeitbuchungAudit {

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        private Long id;

        /** Referenz zur Haupt-Zeitbuchung */
        @Column(name = "zeitbuchung_id", nullable = false)
        private Long zeitbuchungId;

        /** Version dieses Snapshots (1 = initial, 2 = erste Änderung, ...) */
        @Column(nullable = false)
        private Integer version;

        /** Art der Änderung */
        @Enumerated(EnumType.STRING)
        @Column(length = 20, nullable = false)
        private AuditAktion aktion;

        // ============== Snapshot der Daten zum Zeitpunkt der Änderung ==============

        @Column(name = "mitarbeiter_id", nullable = false)
        private Long mitarbeiterId;

        @Column(name = "projekt_id")
        private Long projektId;

        @Column(name = "arbeitsgang_id")
        private Long arbeitsgangId;

        @Column(name = "arbeitsgang_stundensatz_id")
        private Long arbeitsgangStundensatzId;

        @Column(name = "projekt_produktkategorie_id")
        private Long projektProduktkategorieId;

        @Column(name = "start_zeit", nullable = false)
        private LocalDateTime startZeit;

        @Column(name = "ende_zeit")
        private LocalDateTime endeZeit;

        @Column(name = "anzahl_in_stunden", precision = 10, scale = 2)
        private BigDecimal anzahlInStunden;

        @Column(columnDefinition = "TEXT")
        private String notiz;

        // ============== Änderungs-Metadaten ==============

        /** Wer hat die Änderung durchgeführt */
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "geaendert_von_mitarbeiter_id", nullable = false)
        private Mitarbeiter geaendertVon;

        /** Wann wurde die Änderung durchgeführt */
        @Column(name = "geaendert_am", nullable = false)
        private LocalDateTime geaendertAm;

        /** Über welchen Kanal wurde geändert */
        @Enumerated(EnumType.STRING)
        @Column(name = "geaendert_via", length = 50, nullable = false)
        private ErfassungsQuelle geaendertVia;

        /** Begründung für die Änderung (Pflicht bei GEAENDERT/STORNIERT) */
        @Column(name = "aenderungsgrund", columnDefinition = "TEXT")
        private String aenderungsgrund;

        /**
         * Erstellt einen Snapshot aus einer Zeitbuchung.
         */
        public static ZeitbuchungAudit fromZeitbuchung(Zeitbuchung buchung, AuditAktion aktion,
                        Mitarbeiter bearbeiter, ErfassungsQuelle quelle,
                        String grund) {
                ZeitbuchungAudit audit = new ZeitbuchungAudit();
                audit.setZeitbuchungId(buchung.getId());
                audit.setVersion(buchung.getVersion());
                audit.setAktion(aktion);

                // Snapshot der aktuellen Daten
                audit.setMitarbeiterId(buchung.getMitarbeiter().getId());
                // Projekt kann bei PAUSE-Buchungen null sein
                audit.setProjektId(buchung.getProjekt() != null ? buchung.getProjekt().getId() : null);
                audit.setArbeitsgangId(buchung.getArbeitsgang() != null ? buchung.getArbeitsgang().getId() : null);
                audit.setArbeitsgangStundensatzId(buchung.getArbeitsgangStundensatz() != null
                                ? buchung.getArbeitsgangStundensatz().getId()
                                : null);
                audit.setProjektProduktkategorieId(buchung.getProjektProduktkategorie() != null
                                ? buchung.getProjektProduktkategorie().getId()
                                : null);
                audit.setStartZeit(buchung.getStartZeit());
                audit.setEndeZeit(buchung.getEndeZeit());
                audit.setAnzahlInStunden(buchung.getAnzahlInStunden());
                audit.setNotiz(buchung.getNotiz());

                // Metadaten
                audit.setGeaendertVon(bearbeiter);
                audit.setGeaendertAm(LocalDateTime.now());
                audit.setGeaendertVia(quelle);
                audit.setAenderungsgrund(grund);

                return audit;
        }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/ZeitbuchungAudit.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(
    name = "zeitbuchung_audit",
    uniqueConstraints = [
        UniqueConstraint(
            name = "uk_zeitbuchung_audit_version",
            columnNames = ["zeitbuchung_id", "version"],
        ),
    ],
)
open class ZeitbuchungAudit {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @Column(name = "zeitbuchung_id", nullable = false)
    open var zeitbuchungId: Long? = null

    @Column(nullable = false)
    open var version: Int? = null

    @Enumerated(EnumType.STRING)
    @Column(length = 20, nullable = false)
    open var aktion: AuditAktion? = null

    @Column(name = "mitarbeiter_id", nullable = false)
    open var mitarbeiterId: Long? = null

    @Column(name = "projekt_id")
    open var projektId: Long? = null

    @Column(name = "arbeitsgang_id")
    open var arbeitsgangId: Long? = null

    @Column(name = "arbeitsgang_stundensatz_id")
    open var arbeitsgangStundensatzId: Long? = null

    @Column(name = "projekt_produktkategorie_id")
    open var projektProduktkategorieId: Long? = null

    @Column(name = "start_zeit", nullable = false)
    open var startZeit: LocalDateTime? = null

    @Column(name = "ende_zeit")
    open var endeZeit: LocalDateTime? = null

    @Column(name = "anzahl_in_stunden", precision = 10, scale = 2)
    open var anzahlInStunden: BigDecimal? = null

    @Column(columnDefinition = "TEXT")
    open var notiz: String? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "geaendert_von_mitarbeiter_id", nullable = false)
    open var geaendertVon: Mitarbeiter? = null

    @Column(name = "geaendert_am", nullable = false)
    open var geaendertAm: LocalDateTime? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "geaendert_via", length = 50, nullable = false)
    open var geaendertVia: ErfassungsQuelle? = null

    @Column(name = "aenderungsgrund", columnDefinition = "TEXT")
    open var aenderungsgrund: String? = null

    companion object {
        @JvmStatic
        fun fromZeitbuchung(
            buchung: Zeitbuchung,
            aktion: AuditAktion,
            bearbeiter: Mitarbeiter,
            quelle: ErfassungsQuelle,
            grund: String?,
        ): ZeitbuchungAudit {
            val audit = ZeitbuchungAudit()
            audit.zeitbuchungId = buchung.id
            audit.version = buchung.version
            audit.aktion = aktion
            audit.mitarbeiterId = buchung.mitarbeiter?.id
            audit.projektId = buchung.projekt?.longValue("getId")
            audit.arbeitsgangId = buchung.arbeitsgang?.longValue("getId")
            audit.arbeitsgangStundensatzId = buchung.arbeitsgangStundensatz?.longValue("getId")
            audit.projektProduktkategorieId = buchung.projektProduktkategorie?.longValue("getId")
            audit.startZeit = buchung.startZeit
            audit.endeZeit = buchung.endeZeit
            audit.anzahlInStunden = buchung.anzahlInStunden
            audit.notiz = buchung.notiz
            audit.geaendertVon = bearbeiter
            audit.geaendertAm = LocalDateTime.now()
            audit.geaendertVia = quelle
            audit.aenderungsgrund = grund
            return audit
        }

        private fun Any.longValue(getter: String): Long? =
            javaClass.getMethod(getter).invoke(this) as? Long
    }
}
